File: clojure/src/aoc/pathfinding.clj
(ns aoc.pathfinding
  "Pathfinding and search algorithms for Advent of Code."
  (:require [clojure.data.priority-map :refer [priority-map]]))

(defn bfs
  "Breadth-first search from a start node.

   Parameters:
   - start: Starting node
   - neighbors-fn: Function that takes a node and returns neighbors
   - goal?: Predicate function to test if a node is the goal (optional)

   Returns: Map with:
     - :visited - set of all visited nodes
     - :path - path to goal (if goal? provided and found)
     - :distances - map of node to distance from start
     - :parents - map of node to parent node"
  ([start neighbors-fn]
   (bfs start neighbors-fn nil))
  ([start neighbors-fn goal?]
   (loop [queue (conj clojure.lang.PersistentQueue/EMPTY start)
          visited #{start}
          distances {start 0}
          parents {start nil}]
     (if (empty? queue)
       {:visited visited
        :distances distances
        :parents parents}
       (let [current (peek queue)
             queue (pop queue)]
         (if (and goal? (goal? current))
           (let [path (loop [node current
                             path []]
                        (if (nil? node)
                          (reverse path)
                          (recur (parents node) (conj path node))))]
             {:visited visited
              :path path
              :distances distances
              :parents parents
              :goal current})
           (let [current-dist (distances current)
                 neighbors (filter #(not (visited %)) (neighbors-fn current))
                 new-distances (reduce (fn [acc neighbor]
                                         (assoc acc neighbor (inc current-dist)))
                                       distances
                                       neighbors)
                 new-parents (reduce (fn [acc neighbor]
                                       (assoc acc neighbor current))
                                     parents
                                     neighbors)
                 new-visited (into visited neighbors)
                 new-queue (reduce conj queue neighbors)]
             (recur new-queue new-visited new-distances new-parents))))))))

(defn bfs-all-paths
  "Find all nodes reachable from start using BFS.

   Parameters:
   - start: Starting node
   - neighbors-fn: Function that takes a node and returns neighbors

   Returns: Set of all reachable nodes"
  [start neighbors-fn]
  (:visited (bfs start neighbors-fn)))

(defn bfs-shortest-path
  "Find shortest path from start to goal using BFS.

   Parameters:
   - start: Starting node
   - neighbors-fn: Function that takes a node and returns neighbors
   - goal?: Predicate function to test if a node is the goal

   Returns: Vector representing the path, or nil if no path exists"
  [start neighbors-fn goal?]
  (:path (bfs start neighbors-fn goal?)))

(defn dijkstra
  "Dijkstra's algorithm for shortest path with weighted edges.

   Parameters:
   - start: Starting node
   - neighbors-fn: Function that takes a node and returns sequence of [neighbor cost] pairs
   - goal?: Predicate function to test if a node is the goal (optional)

   Returns: Map with:
     - :distances - map of node to minimum distance from start
     - :parents - map of node to parent node in shortest path
     - :path - path to goal (if goal? provided and found)
     - :goal - goal node (if found)"
  ([start neighbors-fn]
   (dijkstra start neighbors-fn nil))
  ([start neighbors-fn goal?]
   (loop [queue (priority-map start 0)
          distances {start 0}
          parents {start nil}
          visited #{}]
     (if (empty? queue)
       {:distances distances
        :parents parents}
       (let [[current current-dist] (peek queue)
             queue (pop queue)]
         (if (visited current)
           (recur queue distances parents visited)
           (if (and goal? (goal? current))
             (let [path (loop [node current
                               path []]
                          (if (nil? node)
                            (reverse path)
                            (recur (parents node) (conj path node))))]
               {:distances distances
                :parents parents
                :path path
                :goal current
                :distance current-dist})
             (let [neighbors (neighbors-fn current)
                   updates (for [[neighbor cost] neighbors
                                 :let [new-dist (+ current-dist cost)
                                       old-dist (get distances neighbor Double/POSITIVE_INFINITY)]
                                 :when (< new-dist old-dist)]
                             [neighbor new-dist])
                   new-distances (reduce (fn [acc [neighbor dist]]
                                           (assoc acc neighbor dist))
                                         distances
                                         updates)
                   new-parents (reduce (fn [acc [neighbor _]]
                                         (assoc acc neighbor current))
                                       parents
                                       updates)
                   new-queue (reduce (fn [q [neighbor dist]]
                                       (assoc q neighbor dist))
                                     queue
                                     updates)]
               (recur new-queue new-distances new-parents (conj visited current))))))))))

(defn dijkstra-shortest-path
  "Find shortest path from start to goal using Dijkstra's algorithm.

   Parameters:
   - start: Starting node
   - neighbors-fn: Function that takes a node and returns sequence of [neighbor cost] pairs
   - goal?: Predicate function to test if a node is the goal

   Returns: Map with :path and :distance, or nil if no path exists"
  [start neighbors-fn goal?]
  (let [result (dijkstra start neighbors-fn goal?)]
    (when (:path result)
      {:path (:path result)
       :distance (:distance result)})))

(defn a-star
  "A* pathfinding algorithm.

   Parameters:
   - start: Starting node
   - neighbors-fn: Function that takes a node and returns sequence of [neighbor cost] pairs
   - heuristic-fn: Function that estimates distance from a node to the goal
   - goal?: Predicate function to test if a node is the goal

   Returns: Map with:
     - :path - path from start to goal
     - :distance - actual cost of the path
     - :goal - goal node"
  [start neighbors-fn heuristic-fn goal?]
  (loop [queue (priority-map start (heuristic-fn start))
         g-scores {start 0}
         parents {start nil}
         visited #{}]
    (if (empty? queue)
      nil
      (let [[current _] (peek queue)
            queue (pop queue)]
        (if (visited current)
          (recur queue g-scores parents visited)
          (if (goal? current)
            (let [path (loop [node current
                              path []]
                         (if (nil? node)
                           (reverse path)
                           (recur (parents node) (conj path node))))]
              {:path path
               :distance (g-scores current)
               :goal current})
            (let [current-g (g-scores current)
                  neighbors (neighbors-fn current)
                  updates (for [[neighbor cost] neighbors
                                :let [tentative-g (+ current-g cost)
                                      old-g (get g-scores neighbor Double/POSITIVE_INFINITY)]
                                :when (< tentative-g old-g)]
                            [neighbor tentative-g])
                  new-g-scores (reduce (fn [acc [neighbor g]]
                                         (assoc acc neighbor g))
                                       g-scores
                                       updates)
                  new-parents (reduce (fn [acc [neighbor _]]
                                        (assoc acc neighbor current))
                                      parents
                                      updates)
                  new-queue (reduce (fn [q [neighbor g]]
                                      (assoc q neighbor (+ g (heuristic-fn neighbor))))
                                    queue
                                    updates)]
              (recur new-queue new-g-scores new-parents (conj visited current)))))))))

(defn flood-fill
  "Flood fill from a starting position.

   Parameters:
   - start: Starting node
   - neighbors-fn: Function that takes a node and returns neighbors

   Returns: Set of all reachable nodes"
  [start neighbors-fn]
  (loop [queue (conj clojure.lang.PersistentQueue/EMPTY start)
         visited #{start}]
    (if (empty? queue)
      visited
      (let [current (peek queue)
            queue (pop queue)
            neighbors (filter #(not (visited %)) (neighbors-fn current))
            new-visited (into visited neighbors)
            new-queue (reduce conj queue neighbors)]
        (recur new-queue new-visited)))))

(defn count-paths
  "Count all paths from start to goal (use with caution - can be exponential!).

   Parameters:
   - start: Starting node
   - neighbors-fn: Function that takes a node and returns neighbors
   - goal?: Predicate function to test if a node is the goal

   Returns: Number of distinct paths"
  [start neighbors-fn goal?]
  (letfn [(count-from [current visited]
            (if (goal? current)
              1
              (let [neighbors (filter #(not (visited %)) (neighbors-fn current))]
                (reduce + 0 (map #(count-from % (conj visited current)) neighbors)))))]
    (count-from start #{})))

File: go/pkg/pathfinding/pathfinding.go
// Package pathfinding provides graph search algorithms like BFS and Dijkstra.
package pathfinding

import (
	"container/heap"
)

// State represents a generic state in a search problem.
// It must be comparable (can be used as a map key).
type State interface {
	comparable
}

// BFS performs breadth-first search to find the shortest path.
// Returns the distance to the goal, or -1 if no path exists.
//
// Parameters:
//   - start: the starting state
//   - goal: a function that returns true if the state is a goal
//   - neighbors: a function that returns all neighbors of a state
func BFS[S State](start S, goal func(S) bool, neighbors func(S) []S) int {
	if goal(start) {
		return 0
	}

	visited := make(map[S]bool)
	queue := []S{start}
	visited[start] = true
	distance := 0

	for len(queue) > 0 {
		distance++
		size := len(queue)

		for i := 0; i < size; i++ {
			current := queue[0]
			queue = queue[1:]

			for _, next := range neighbors(current) {
				if visited[next] {
					continue
				}
				if goal(next) {
					return distance
				}
				visited[next] = true
				queue = append(queue, next)
			}
		}
	}

	return -1
}

// BFSPath performs breadth-first search and returns the path to the goal.
// Returns the path as a slice of states, or nil if no path exists.
func BFSPath[S State](start S, goal func(S) bool, neighbors func(S) []S) []S {
	if goal(start) {
		return []S{start}
	}

	visited := make(map[S]bool)
	parent := make(map[S]S)
	queue := []S{start}
	visited[start] = true

	var endState S
	found := false

	for len(queue) > 0 && !found {
		current := queue[0]
		queue = queue[1:]

		for _, next := range neighbors(current) {
			if visited[next] {
				continue
			}
			visited[next] = true
			parent[next] = current

			if goal(next) {
				endState = next
				found = true
				break
			}
			queue = append(queue, next)
		}
	}

	if !found {
		return nil
	}

	// Reconstruct path
	var path []S
	for state := endState; ; state = parent[state] {
		path = append(path, state)
		if state == start {
			break
		}
	}

	// Reverse path
	for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
		path[i], path[j] = path[j], path[i]
	}

	return path
}

// BFSAll performs breadth-first search and visits all reachable states.
// Returns a map of states to their distances from the start.
func BFSAll[S State](start S, neighbors func(S) []S) map[S]int {
	distances := make(map[S]int)
	queue := []S{start}
	distances[start] = 0

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		currentDist := distances[current]

		for _, next := range neighbors(current) {
			if _, visited := distances[next]; visited {
				continue
			}
			distances[next] = currentDist + 1
			queue = append(queue, next)
		}
	}

	return distances
}

// Edge represents a weighted edge to a neighbor state.
type Edge[S State] struct {
	State S
	Cost  int
}

// priorityQueueItem is an item in the priority queue.
type priorityQueueItem[S State] struct {
	state    S
	priority int
	index    int
}

// priorityQueue implements heap.Interface.
type priorityQueue[S State] []*priorityQueueItem[S]

func (pq priorityQueue[S]) Len() int { return len(pq) }

func (pq priorityQueue[S]) Less(i, j int) bool {
	return pq[i].priority < pq[j].priority
}

func (pq priorityQueue[S]) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *priorityQueue[S]) Push(x interface{}) {
	n := len(*pq)
	item := x.(*priorityQueueItem[S])
	item.index = n
	*pq = append(*pq, item)
}

func (pq *priorityQueue[S]) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil
	item.index = -1
	*pq = old[0 : n-1]
	return item
}

// Dijkstra performs Dijkstra's algorithm to find the shortest weighted path.
// Returns the minimum cost to reach the goal, or -1 if no path exists.
//
// Parameters:
//   - start: the starting state
//   - goal: a function that returns true if the state is a goal
//   - neighbors: a function that returns weighted edges to neighbors
func Dijkstra[S State](start S, goal func(S) bool, neighbors func(S) []Edge[S]) int {
	if goal(start) {
		return 0
	}

	dist := make(map[S]int)
	pq := make(priorityQueue[S], 0)
	heap.Init(&pq)

	dist[start] = 0
	heap.Push(&pq, &priorityQueueItem[S]{state: start, priority: 0})

	for pq.Len() > 0 {
		item := heap.Pop(&pq).(*priorityQueueItem[S])
		current := item.state
		currentDist := item.priority

		// Skip if we've found a better path
		if d, ok := dist[current]; ok && currentDist > d {
			continue
		}

		if goal(current) {
			return currentDist
		}

		for _, edge := range neighbors(current) {
			newDist := currentDist + edge.Cost
			if oldDist, ok := dist[edge.State]; !ok || newDist < oldDist {
				dist[edge.State] = newDist
				heap.Push(&pq, &priorityQueueItem[S]{state: edge.State, priority: newDist})
			}
		}
	}

	return -1
}

// DijkstraAll performs Dijkstra's algorithm and returns distances to all reachable states.
// Returns a map of states to their minimum costs from the start.
func DijkstraAll[S State](start S, neighbors func(S) []Edge[S]) map[S]int {
	dist := make(map[S]int)
	pq := make(priorityQueue[S], 0)
	heap.Init(&pq)

	dist[start] = 0
	heap.Push(&pq, &priorityQueueItem[S]{state: start, priority: 0})

	for pq.Len() > 0 {
		item := heap.Pop(&pq).(*priorityQueueItem[S])
		current := item.state
		currentDist := item.priority

		// Skip if we've found a better path
		if d, ok := dist[current]; ok && currentDist > d {
			continue
		}

		for _, edge := range neighbors(current) {
			newDist := currentDist + edge.Cost
			if oldDist, ok := dist[edge.State]; !ok || newDist < oldDist {
				dist[edge.State] = newDist
				heap.Push(&pq, &priorityQueueItem[S]{state: edge.State, priority: newDist})
			}
		}
	}

	return dist
}

// DijkstraPath performs Dijkstra's algorithm and returns the path to the goal.
// Returns the path as a slice of states, or nil if no path exists.
func DijkstraPath[S State](start S, goal func(S) bool, neighbors func(S) []Edge[S]) []S {
	if goal(start) {
		return []S{start}
	}

	dist := make(map[S]int)
	parent := make(map[S]S)
	pq := make(priorityQueue[S], 0)
	heap.Init(&pq)

	dist[start] = 0
	heap.Push(&pq, &priorityQueueItem[S]{state: start, priority: 0})

	var endState S
	found := false

	for pq.Len() > 0 && !found {
		item := heap.Pop(&pq).(*priorityQueueItem[S])
		current := item.state
		currentDist := item.priority

		// Skip if we've found a better path
		if d, ok := dist[current]; ok && currentDist > d {
			continue
		}

		if goal(current) {
			endState = current
			found = true
			break
		}

		for _, edge := range neighbors(current) {
			newDist := currentDist + edge.Cost
			if oldDist, ok := dist[edge.State]; !ok || newDist < oldDist {
				dist[edge.State] = newDist
				parent[edge.State] = current
				heap.Push(&pq, &priorityQueueItem[S]{state: edge.State, priority: newDist})
			}
		}
	}

	if !found {
		return nil
	}

	// Reconstruct path
	var path []S
	for state := endState; ; state = parent[state] {
		path = append(path, state)
		if state == start {
			break
		}
	}

	// Reverse path
	for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
		path[i], path[j] = path[j], path[i]
	}

	return path
}
